(WIP) expose effective downlink + RTT

igrigorik opened the pull request:
This is a first run at integrating the JS interfaces of the ECT proposal discussed in https://github.com/WICG/netinfo/issues/46#issuecomment-280165405. 

- The definitions for `downlink` and `rtt` need more work: we don't want to over-specify how these are determined by the UA, but we do need to provide some explanation and guidance for what kind of signals are accounted for.
- Related to above, we should also consider if we want to provide some guidance on thresholds for triggering onchange updates - related https://github.com/WICG/netinfo/issues/30.

/cc @tarunban @bmaurer @yoavweiss @marcoscaceres ptal, would love to hear your feedback.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
[Preview](https://s3.amazonaws.com/pr-preview/WICG/netinfo/ect.html) | [Diff](https://s3.amazonaws.com/pr-preview/WICG/netinfo/c332520...1afd51e.html)

igrigorik committed:
expose effective downlink + RTT

tarunban commented on index.html:
```diff
@@ -111,32 +111,82 @@
+    </thead>
+    <tbody>
+      <tr>
+        <td><dfn>slow2g</dfn></td>
```
Probably it does not matter, but we are now using slow-2g everywhere.
s/slow2g/slow-2g/

marcoscaceres replied:
actually, we should change this one to align with the rest of the platform.

tarunban commented on index.html:
```diff
@@ -152,32 +202,46 @@
 };
 
 typedef unrestricted double Megabit;
+typedef unsigned long long Millisecond;
```
s/long long/long/

marcoscaceres replied:
@igrigorik make sure this matches whatever is used in other specs that rely on millis.

marcoscaceres committed:
fix markdown and define 'onchange'

marcoscaceres commented:
Unfortunately, fixing the table issue took a little longer than I was expecting. Will have to review this more fully tomorrow.

marcoscaceres commented on index.html:
```diff
@@ -111,32 +111,82 @@
+      </tr>
+    </tbody>
+  </table>
+  <p class="note">The above thresholds are generated from real user observations by Chrome over different connection types and network technologies (e.g., EDGE, GPRS).</p>
```
Is there any public document that describes the observations made by the chrome team? That is, something we can cite?

marcoscaceres committed:
editorial: typo

marcoscaceres committed:
fix: bad xrefs

marcoscaceres reviewed (approved):
This LGTM. 

I've fixed a few editorial things and broken links.

marcoscaceres committed:
fix: enums can't have trailing commas

marcoscaceres committed:
fix: s/slow2g/slow-2g

igrigorik committed:
explain thresholds + ect algorithm

- added threshold explanations in note
- added ECT update to the onchange algorithm

igrigorik commented:
@marcoscaceres updated the note with additional context on how the data was obtained. We don't have a publication we can point to, at least today -- this will be the source of truth. @tarunban do we want to make any points about updating these values in the future, or not?

Also, a couple additional fixes:

- added ECT to the onchange algorithm.
- made a note that RTT and BW values should be rounded to 25 ms / kb respectively.

Anything else we need to cover here?

marcoscaceres committed:
refactor: link things a little better

marcoscaceres commented on index.html:
```diff
@@ -385,7 +394,7 @@
   ## Privacy Considerations
 
-  The Network Information API exposes information about the first network hop between the user agent and the server; specifically, the type of connection and the upper bound of the downlink speed, as well as signals whenever this information changes. Such information may be used to:
+  The Network Information API exposes information about the observed network bandwidth and latency and the first network hop between the user agent and the server; specifically, the type of connection and the upper bound of the downlink speed, as well as signals whenever this information changes. Such information may be used to:
```
nit: and ... and.

igrigorik replied:
Good catch, will fix.

marcoscaceres reviewed (approved):
Just minor nit.

igrigorik committed:
convert thresholds to normative text

igrigorik committed:
fix ms rounding

yoavweiss commented:
LGTM!!

igrigorik committed:
collapse extra and

igrigorik commented:
There's probably some editorial improvements to be made still, but I think we have a solid v1. I'll merge this, and let's iterate on it in followup PRs.

@marcoscaceres thanks for help with fixing up all the respec linking and formatting!
